### PR TITLE
bugfix error importing past backups w/o loan dates

### PR DESCRIPTION
**Pull request by beradeep**
## Pull Request (PR) Checklist
Please check if your pull request fulfills the following requirements:
- [x] The PR is submitted to the `main` branch.
- [x] I've read the [Contribution Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/CONTRIBUTING.md) and my PR doesn't break the "Contributing Rules".
- [x] I've read the [Architecture Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/Architecture.md).
- [x] I confirm that I've run the code locally and everything works as expected.
- [ ] 🎬 I've attached a **screen recoding** of the changes. 

Describe with a few bullets **what's new:**
- Fixed backwards db compatibility.
- Fixed errors when importing loans without dates from older app versions.

### CODE REVIEW

**Review by ILIYANGERMANOV — CHANGES_REQUESTED** [review]
Thank you for the fix! LGTM but I believe you need to make a new DB migration for that because the app will crash for users who have installed the app with the Loan dates addition. Do it this change as a separate DB migration `124 to 125`

**Comment by beradeep** [comment]
Yes. I should have done it that way.

File: ivy-data/src/main/java/com/ivy/data/db/migration/Migration124to125_LoanEditDateTime.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import androidx.room.migration.Migration
+import androidx.sqlite.db.SupportSQLiteDatabase
+
+class Migration124to125_LoanEditDateTime : Migration(124, 125) {
+    companion object {
+        private const val LOANS_TABLE = "loans"
+        private const val LOANS_NEW_TABLE = "loans_new"
+    }
+
+    override fun migrate(db: SupportSQLiteDatabase) {
```
Comment (ILIYANGERMANOV) [inline ivy-data/src/main/java/com/ivy/data/db/migration/Migration124to125_LoanEditDateTime.kt:12]: Ahh, this migration seems complicated and risky. Is there a better way?

**Review by ILIYANGERMANOV — COMMENTED** [review]
Looks better @beradeep but I still have some concerns about the db migration. Also "loans_v2" is a better naming for the table instead of "_new"

File: ivy-data/src/main/java/com/ivy/data/db/entity/LoanEntity.kt [hunk 2]
```diff
@@ -37,7 +37,7 @@ data class LoanEntity(
     val isDeleted: Boolean = false,
     @SerialName("dateTime")
     @Serializable(with = KSerializerLocalDateTime::class)
-    val dateTime: LocalDateTime,
+    val dateTime: LocalDateTime? = null,
```
Comment (ILIYANGERMANOV) [inline ivy-data/src/main/java/com/ivy/data/db/entity/LoanEntity.kt:40]: If we're using "loans_new" this table must also be renamed. The migration overall seems very.risky and I doubt it works in all cases

File: ivy-data/src/main/java/com/ivy/data/db/migration/Migration124to125_LoanEditDateTime.kt (hunk 1 above)
Comment (beradeep) [reply ivy-data/src/main/java/com/ivy/data/db/migration/Migration124to125_LoanEditDateTime.kt:12]: Why though? SQLite doesn't support editing column datatypes, unlike SQL. And the table name remains "loans" as previous, as it gets renamed just after copying the existing data from the old table.

File: ivy-data/src/main/java/com/ivy/data/db/entity/LoanEntity.kt (hunk 2 above)
Comment (beradeep) [reply ivy-data/src/main/java/com/ivy/data/db/entity/LoanEntity.kt:40]: "loans_new" is just for creating a new table. The new table is replacing the old one with the same name "loans", at the end of the migration.

Comment (ILIYANGERMANOV) [reply ivy-data/src/main/java/com/ivy/data/db/entity/LoanEntity.kt:40]: Got it - I'll test it. Revert `124.json` to be like the one in main

Comment (beradeep) [reply ivy-data/src/main/java/com/ivy/data/db/entity/LoanEntity.kt:40]: > Got it - I'll test it. Revert `124.json` to be like the one in main

I am a bit confused why should we revert `124.json`?

**Review by ILIYANGERMANOV — CHANGES_REQUESTED** [review]
The migration code isn't working, see the attached video:


https://github.com/Ivy-Apps/ivy-wallet/assets/5564499/b663aed9-8827-40d9-8f10-2befcbbe8531

**Comment by beradeep** [comment]
> The migration code isn't working, see the attached video:
> 
>  Screen_Recording_20240116_200746_One.UI.Home.mp4

It's weird. It worked for me though. Did it crash when you opened loans?

**Comment by ILIYANGERMANOV** [comment]
> > The migration code isn't working, see the attached video:
> > Screen_Recording_20240116_200746_One.UI.Home.mp4
> 
> It's weird. It worked for me though. Did it crash when you opened loans?

Yes, exactly. The migration is executed when you open the Loans tab. Also you must have loans from the previous APK

**Comment by beradeep** [comment]
> > > The migration code isn't working, see the attached video:
> > > Screen_Recording_20240116_200746_One.UI.Home.mp4
> > 
> > 
> > It's weird. It worked for me though. Did it crash when you opened loans?
> 
> Yes, exactly. The migration is executed when you open the Loans tab. Also you must have loans from the previous APK

what were the logs?

**Review by ILIYANGERMANOV — APPROVED** [review]
LGTM! Haven't tested the migration but let's hope that it works

**Pull request merged**